File: crates/baro-tui/src/events.rs
#![allow(dead_code)]

use serde::Deserialize;

#[derive(Debug, Deserialize, Clone)]
pub struct StoryInfo {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub depends_on: Vec<String>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct DagNode {
    pub id: String,
    pub title: String,
}

#[derive(Debug, Deserialize, Clone)]
pub struct DoneStats {
    pub stories_completed: u32,
    pub stories_skipped: u32,
    pub total_commits: u32,
    pub files_created: u32,
    pub files_modified: u32,
}

#[derive(Debug, Deserialize, Clone)]
#[serde(tag = "type")]
pub enum BaroEvent {
    #[serde(rename = "init")]
    Init {
        project: String,
        stories: Vec<StoryInfo>,
    },

    #[serde(rename = "dag")]
    Dag {
        levels: Vec<Vec<DagNode>>,
    },

    #[serde(rename = "story_start")]
    StoryStart {
        id: String,
        title: String,
    },

    #[serde(rename = "story_log")]
    StoryLog {
        id: String,
        line: String,
    },

    #[serde(rename = "story_complete")]
    StoryComplete {
        id: String,
        duration_secs: u64,
        files_created: u32,
        files_modified: u32,
    },

    #[serde(rename = "story_error")]
    StoryError {
        id: String,
        error: String,
        attempt: u32,
        max_retries: u32,
    },

    #[serde(rename = "story_retry")]
    StoryRetry {
        id: String,
        attempt: u32,
    },

    #[serde(rename = "progress")]
    Progress {
        completed: u32,
        total: u32,
        percentage: u32,
    },

    #[serde(rename = "done")]
    Done {
        total_time_secs: u64,
        stats: DoneStats,
    },
}

File: packages/baro-app/src/core/events.ts
import type { Writable } from "stream"

export interface StoryInfo {
    id: string
    title: string
    dependsOn: string[]
}

export interface DoneStats {
    stories_completed: number
    stories_skipped: number
    total_commits: number
    files_created: number
    files_modified: number
}

export type BaroEvent =
    | { type: "init"; project: string; stories: StoryInfo[] }
    | { type: "dag"; levels: { id: string; title: string }[][] }
    | { type: "story_start"; id: string; title: string }
    | { type: "story_log"; id: string; line: string }
    | {
          type: "story_complete"
          id: string
          duration_secs: number
          files_created: number
          files_modified: number
      }
    | { type: "story_error"; id: string; error: string; attempt: number; max_retries: number }
    | { type: "story_retry"; id: string; attempt: number }
    | { type: "progress"; completed: number; total: number; percentage: number }
    | { type: "push_status"; id: string; success: boolean; error?: string }
    | { type: "done"; total_time_secs: number; stats: DoneStats }

export class BaroEventEmitter {
    constructor(private stream: Writable = process.stdout) {}

    emit(event: BaroEvent): void {
        this.stream.write(JSON.stringify(event) + "\n")
    }
}
